http_service/bugbug_http/utils.py

```diff
@@ -30,3 +36,91 @@ def get_hgmo_stack(branch: str, revision: str) -> list:
     r = requests.get(url)
     r.raise_for_status()
     return r.json()["changesets"]
+
+
+# A simple TTL cache to use with models. Because we expect the number of models
+# in the service to not be very large, simplicity of implementation is
+# preferred to algorithmic efficiency of operations.
+#
+# Called an 'Idle' TTL cache because TTL of items is reset after every get
+Key = TypeVar("Key")
+Value = TypeVar("Value")
+
+
+class IdleTTLCache(Generic[Key, Value]):
+    def __init__(self, ttl: timedelta):
+        self.ttl = ttl
+        self.items_last_touched: Dict[Key, datetime.datetime] = {}
+        self.items: Dict[Key, Value] = {}
+
+    def __setitem__(self, key, item):
+        self.items[key] = item
+        self.items_last_touched[key] = datetime.datetime.now()
+
+    def __getitem__(self, key):
+        item = self.items[key]
+        self.items_last_touched[key] = datetime.datetime.now()
+        return item
+
+    def __contains__(self, key):
+        return key in self.items
+
+    def purge_expired_entries(self):
+        purge_entries_before = datetime.datetime.now() - self.ttl
+        for (key, time_last_touched) in list(self.items_last_touched.items()):
+            if time_last_touched < purge_entries_before:
+                del self.items_last_touched[key]
+                del self.items[key]
+
+    def start_ttl_thread(self):
+        self.purge_expired_entries()
+        timer = threading.Timer(self.ttl.total_seconds(), self.start_ttl_thread)
+        timer.setDaemon(True)
+        timer.start()
+
+
+class ReadthroughTTLCache(Generic[Key, Value]):
+    def __init__(self, ttl: timedelta, load_item_function: Callable[[Key], Value]):
+        self.ttl = ttl
+        self.load_item_function = load_item_function
+        self.items_last_accessed: Dict[Key, datetime.datetime] = {}
+        self.items_storage: Dict[Key, Value] = {}
+
+    def __contains__(self, key):
+        return key in self.items_storage
+
+    def get(self, key):
+        item = None
+        if item in self.items_storage:
+            item = self.items_storage[key]
+        else:
+            item = self.load_item_function(key)
+            # Cache the item only if it was last accessed within the past TTL seconds
+            # Note that all entries in items_last_accessed are purged if item was not
+            # accessed in the last TTL seconds.
+            if key in self.items_last_accessed:
+                self.items_storage[key] = item
+
+        self.items_last_accessed[key] = datetime.datetime.now()
+        return item
+
+    def force_store(self, key):
```

Comment:
I'd say we can remove this function, in the tests you can just `get` twice to force a store

Reply:
Below is my reasoning for having force_store - its subjective, and I'm happy to switch to using two gets if you prefer, but at this time I'd like to try to convince you that the force_store function belongs here, because in my eyes its better to have it than use two gets for the following reasons:

1. This is more future proof. If in the future we want to change the internals of the cache to store models after, say. three accesses in the past hour, we won't need to update as many unit tests.

2. I think its more readable this way. I originally had two gets in my unit tests, but upon seeing those two gets sitting next to each other I felt like someone who wasn't familiar with the code might be confused by what was happening and be compelled to read the cache internals to understand the test, so I wrote a comment above each test. 

Thus, each place where we now have a force_store was replaced by a lot of characters - two calls to get, and a comment. The codebase ended up being as clear, and less verbose in my eyes when I had the force_store function.

Reply:
OK, you convinced me with the first reason ;)
Maybe we could add a `force_store` parameter to `get` though, so we don't need to have a separate function.

Reply:
Will update get to take a parameter

Reply:
and done!